### PR TITLE
feat: ETH TRACE API: Trace installed bytecode in the Ethereum JSON-RPC trace_block output

### PR DESCRIPTION
**Pull request by aarshkshah1992**
## Related Issues
Closes https://github.com/filecoin-project/lotus/issues/11635.

## Checklist

Before you mark the PR ready for review, please make sure that:

- [ ] Commits have a clear commit message.
- [ ] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [ ] If the PR affects users (e.g., new feature, bug fix, system requirements change), update the CHANGELOG.md and add details to the UNRELEASED section.
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [ ] CI is green

### CODE REVIEW

**Comment by aarshkshah1992** [comment]
@Stebalien Does this look correct ?

File: node/impl/full/eth_trace.go [hunk 1]
```diff
@@ -498,6 +499,13 @@ func traceEthCreate(env *environment, addr []int, et *types.ExecutionTrace) (*et
 		// So we don't try and include a sentinel "impossible bytecode"
 		// value (the value specified by EIP-3541).
 		output = []byte{0xFE}
+		if deployedCode != nil {
+			var getBytecodeReturn evm12.GetBytecodeReturn
+			if err := getBytecodeReturn.UnmarshalCBOR(bytes.NewReader(et.MsgRct.Return)); err == nil && getBytecodeReturn.Cid != nil {
```
Comment (aarshkshah1992) [inline node/impl/full/eth_trace.go:504]: @Stebalien 

Is this correct ? Can I get the byte code from the return value of the receipt on the top level transaction ?

Comment (aarshkshah1992) [reply node/impl/full/eth_trace.go:504]: How else can I get the `cid` of the code here ?

File: node/impl/full/eth_trace.go [hunk 2]
```diff
@@ -498,6 +499,13 @@ func traceEthCreate(env *environment, addr []int, et *types.ExecutionTrace) (*et
...
+				deployedCode[abi.ActorID(id)] = getBytecodeReturn.Cid
```
Comment (aarshkshah1992) [inline node/impl/full/eth_trace.go:505]: @Stebalien This is not correct. How do I get the byte code of the contract here ?

Comment (aarshkshah1992) [reply node/impl/full/eth_trace.go:505]: Do I need to make a call to the `GetBytecode` method of the contract Actor ?
Looks like none of the trace types we here have the byte code in them.

**Comment by jennijuju** [comment]
cc @eshon this PR should fix the issue you brought up in our 1:1!

File: node/impl/full/eth.go [hunk 3]
```diff
@@ -888,7 +890,12 @@ func (a *EthModule) EthTraceBlock(ctx context.Context, blkNum string) ([]*ethtyp
 			return nil, xerrors.Errorf("when processing message %s: %w", ir.MsgCid, err)
 		}
 
-		err = buildTraces(env, []int{}, &ir.ExecutionTrace)
+		var deployed map[abi.ActorID]*cid.Cid
+		if ir.MsgRct.ExitCode.IsSuccess() {
+			deployed = deployedCode
+		}
```
Comment (Stebalien) [inline node/impl/full/eth.go:896]: We need to do this per sub-trace. The top-level message can succeed while sub-messages cain fail.

Basically, move this into `buildTraces` and, if the receipt of the currently processed message has a non-zero exit code, set this to nil for the sub-call.

File: node/impl/full/eth.go [hunk 4]
```diff
@@ -904,9 +911,62 @@ func (a *EthModule) EthTraceBlock(ctx context.Context, blkNum string) ([]*ethtyp
...
+		return nil, xerrors.Errorf("failed to update contract addresses: %w", err)
+	}
+
 	return allTraces, nil
 }
 
+func (a *EthModule) updateContractAddresses(ctx context.Context, ts *types.TipSet, allTraces []*ethtypes.EthTrace, deployedCode map[abi.ActorID]*cid.Cid) error {
+	// Update Contract Addresses if contract still exists
+	for _, trace := range allTraces {
+		switch r := trace.Result.(type) {
```
Comment (Stebalien) [inline node/impl/full/eth.go:930]: This isn't correct, we want to update the last one _only_. The idea was to record a pointer to `r.Code` in `deployedCode`, so we could set it without even having to look at the traces. That is:

```go
deployedCode := make(map[abi.ActorID]*EthBytes)
```

Then, inside the `buildTraces` logic, we'd record `deployedCode[receiverID] = &r.Code`.

Finally, we'd iterate over `deployedCode` and set `*deployedCode[receiverID] = loadByteCode(receiverID)`, as long as the receiver is still live.

File: node/impl/full/eth_trace.go (hunk 2 above)
Comment (Stebalien) [reply node/impl/full/eth_trace.go:505]: Sorry, I missed this question. You can actually just read the bytecode from the EVM actor state (load the actor and call `GetBytecode` using the actor shims.

However, you have to do this _after_ building the trace (see https://github.com/filecoin-project/lotus/pull/11984#discussion_r1612060558)

**Comment by BigLep** [comment]
@aarshkshah1992 : I'm moving this back to draft given it hasn't been updated for a few months and I'm not aware of it being a priority currently.

**Comment by aarshkshah1992** [comment]
@BigLep Closing this as Im not sure if this is correct and it needs engagement from Steb if we get to it again.

**Comment by Stebalien** [comment]
I've restored the branch here because the code may be useful in the future.

**Pull request closed without merging**